### PR TITLE
Avoid out-of-bound access when setting shape dimension size

### PR DESCRIPTION
**Pull request by antiagainst**
The number of implicit loops in a Linalg op can be greater than
the shape rank of the result. We can tile loops not corresponding
to the result dimension. Guard against such cases.

### CODE REVIEW

File: iree/compiler/Codegen/Utils/Utils.cpp [hunk 1]
```diff
@@ -140,6 +140,7 @@ SmallVector<int64_t> getUntiledResultShape(linalg::LinalgOp linalgOp,
 
   for (auto pair : llvm::zip(llvm::reverse(partitionedLoops), loopInfo)) {
     unsigned loopIndex = std::get<0>(pair);
+    if (loopIndex >= outputShape.size()) continue;
```
Comment (ThomasRaoux) [inline iree/compiler/Codegen/Utils/Utils.cpp:143]: In what case does this happen? I would think the output rank matches the iteration space? Maybe I'm not missing something about this logic.

Comment (antiagainst) [reply iree/compiler/Codegen/Utils/Utils.cpp]: Not necessary. For example, for `linalg.nhwc`, we have `(N, OH, OW, KH, KW, C)` but it's output only have `(N, OH, OW, C)`. If we just use the `loopIndex` for `C` (which is 5) to access the output shape, it would be OOB.

But on a second thought, this is actually not the proper fix. The real issue is that it's assuming the first iterators of the Linalg op are always used to access output shape dimensions. That's not true; instead we need to figure out which loops are used to access output shape dimensions. I've fixed this in a second commit. Thanks for spotting this!

File: iree/compiler/Codegen/SPIRV/test/config_default_linalg_ops.mlir [hunk 2]
```diff
@@ -280,3 +280,79 @@ hal.executable @avg_pool {
...
+            %8 = affine.min affine_map<(d0)[s0] -> (s0, -d0 + 10)>(%arg1)[%workgroup_size_x]
+            %9 = flow.dispatch.tensor.load %0, offsets = [%arg0, %arg1], sizes = [%7, %8], strides = [1, 1] : !flow.dispatch.tensor<readonly:1x10xf32> -> tensor<?x?xf32>
+            %10 = flow.dispatch.tensor.load %1, offsets = [%arg1], sizes = [%8], strides = [1] : !flow.dispatch.tensor<readonly:10xf32> -> tensor<?xf32>
+            %11 = linalg.init_tensor [%8] : tensor<?xf32>
+            %12 = linalg.generic {
+              indexing_maps = [
+                affine_map<(d0, d1) -> (d0, d1)>,
+                affine_map<(d0, d1) -> (d1)>,
+                affine_map<(d0, d1) -> (d1)>],
+              iterator_types = ["parallel", "parallel"]
```
Comment (ThomasRaoux) [inline iree/compiler/Codegen/SPIRV/test/config_default_linalg_ops.mlir:330]: Is this correct? If there are two parallel iterator I would expect the output affine map to have two results. Looking at the affine map this would be a reduction?

File: iree/compiler/Codegen/Utils/Utils.cpp (hunk 1 above)
Comment (ThomasRaoux) [reply iree/compiler/Codegen/Utils/Utils.cpp]: In this case KH and KW would be reduction dimensions and wouldn't be part of `partitionedLoops`, right? I would expect all the loops in partiotionedLoops to be parallel loop and match to an output dimension but I could be missing something.

Comment (antiagainst) [reply iree/compiler/Codegen/Utils/Utils.cpp]: Yes, `KW` and `KH` won't be partitioned (so not in `partitionedLoops`) but `C` can be in it. The `loopIndex` for `C` is 5, which is greater than the output shape's rank. Accessing `outputShape` with `loopIndex` can see OOB for such case.

File: iree/compiler/Codegen/SPIRV/test/config_default_linalg_ops.mlir (hunk 2 above)
Comment (antiagainst) [reply iree/compiler/Codegen/SPIRV/test/config_default_linalg_ops.mlir:330]: I'm also a bit puzzled by this test case actually. It's generated in real from mnist model. Though it can still have wrong semantics. Maybe it's a corner case where we have size-1 dim (effectively can be treated either as parallel or reduction)?

File: iree/compiler/Codegen/Utils/Utils.cpp (hunk 1 above)
Comment (antiagainst) [reply iree/compiler/Codegen/Utils/Utils.cpp]: For pooling, we normally want to distribute along `OH`, `OW`, and `C`. `N` is typically one; no need to distribute while `C` can be quite large.

File: iree/compiler/Codegen/Utils/Utils.cpp [hunk 3]
```diff
@@ -138,12 +138,32 @@ SmallVector<int64_t> getUntiledResultShape(linalg::LinalgOp linalgOp,
...
+
   for (auto pair : llvm::zip(llvm::reverse(partitionedLoops), loopInfo)) {
     unsigned loopIndex = std::get<0>(pair);
     const LoopTilingAndDistributionInfo &loopInfo = std::get<1>(pair);
+    // If we know the static upper bound of this loop..
     if (Optional<int64_t> attrValue =
             getConstantIntValue(loopInfo.untiledUpperBound)) {
-      outputShape[loopIndex] = *attrValue;
+      // ..and it accesses one output dimension..
+      if (loopToOutputDimMap.count(loopIndex)) {
```
Comment (ThomasRaoux) [inline iree/compiler/Codegen/Utils/Utils.cpp:163]: nit: I wonder if this should be an assert instead. This should always be true right?

**Review by ThomasRaoux — APPROVED** [review]
That looks good to me. Thanks for explaining. I'm a bit curious about the linalg op, I'll ask around :)

**Comment by MaheshRavishankar** [comment]
The issue is so far we have been distributing loops without paying attention to static values of shapes (on purpose). If we want to be smarter about this, we need to ensure that 
1) The dispatch region formation in flow does not distribute unit-dims.
2) The executable translation knows about this.

We dont have a fixed mechanism to propagate such information. In absence of that so far it is done by keep this method https://github.com/google/iree/blob/02cfcd1e5561c0db0470d09644a1f527c259c208/iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp#L653 and https://github.com/google/iree/blob/02cfcd1e5561c0db0470d09644a1f527c259c208/iree/compiler/Codegen/Utils/Utils.cpp#L58 in sync.

The correct way to fix this is add an interface that allows you to specify which operations are distributed how, and we use that interface consistently at the Flow level and at the HAL level. 

Any takers for adding this interface. It should be fairly simple to do this.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
I dont understand the logic here, and it seems too complicated. I left a comment on the use of interface to be consistent with what loops are partitioned. That just seems to be the way to fix it. This looks fairly ad-hoc to me.

**Comment by antiagainst** [comment]
Agreed that distributing unit dim is bad (esp. for performance). I'm happy to look at and add an interface to allow us avoid distributing unit dims. However, it's disjoint from the bug this patch is meant to solve. This patch is meant to address an OOB access bug, which will reveal even if we don't distribute unit dimensions. Let me explain more:

Right now the pooling op definition in Linalg is:

https://github.com/llvm/llvm-project/blob/f67171ac5896d232b4b9826937ce97064909770b/mlir/python/mlir/dialects/linalg/opdsl/ops/core_named_ops.py#L411-L426

For its output, the indexing map looks like:

https://github.com/llvm/llvm-project/blob/f67171ac5896d232b4b9826937ce97064909770b/mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOps.yaml#L1913-L1914

So we have 6 implicit loops `(d0, d1, d2, d3, d4, d5)`, and we index into the output with `(d0, d1, d2, d5)` (which corresponds to `(N, OH, OW, C)`).

Say we omit `N` (which is typically 1) and distribute along `OH`, `OW`, and `C` (they are typically not 1). (This is actually  how we distribute for pooling ops actually right now in IREE.) Then the `partitionedLoops` in the following current impl:

https://github.com/google/iree/blob/02cfcd1e5561c0db0470d09644a1f527c259c208/iree/compiler/Codegen/Utils/Utils.cpp#L134-L148

The `partitionedLoops` would return `[1, 2, 5]`. Then at L146 above, we could use `5` to index into `outputShape`, which only have 4 dimensions. That's OOB.

The real issue is that the current impl is assuming the first batch of implicit loops are used to indexing into the output. So the fix we need is to add the map from `loopToOutputDimMap` to break that false assumption. Avoid distributing along unit dims is something we certainly want, but it won't fix the OOB issue here. We also need the fix here. :)

I'm using the illformed `linalg.generic` as the test instead of using `linalg.pool` because we don't really have configurations for pooling ops as root to trigger this OOB; don't want to just add a configuration for this fix, as that has performance implications. 

Does this make sense? Happy to explain more if not.

**Comment by MaheshRavishankar** [comment]
> Agreed that distributing unit dim is bad (esp. for performance). I'm happy to look at and add an interface to allow us avoid distributing unit dims. However, it's disjoint from the bug this patch is meant to solve. This patch is meant to address an OOB access bug, which will reveal even if we don't distribute unit dimensions. Let me explain more:
> 
> Right now the pooling op definition in Linalg is:
> 
> [llvm/llvm-project@`f67171a`/mlir/python/mlir/dialects/linalg/opdsl/ops/core_named_ops.py#L411-L426](https://github.com/llvm/llvm-project/blob/f67171ac5896d232b4b9826937ce97064909770b/mlir/python/mlir/dialects/linalg/opdsl/ops/core_named_ops.py#L411-L426)
> 
> For its output, the indexing map looks like:
> 
> [llvm/llvm-project@`f67171a`/mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOps.yaml#L1913-L1914](https://github.com/llvm/llvm-project/blob/f67171ac5896d232b4b9826937ce97064909770b/mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOps.yaml#L1913-L1914)
> 
> So we have 6 implicit loops `(d0, d1, d2, d3, d4, d5)`, and we index into the output with `(d0, d1, d2, d5)` (which corresponds to `(N, OH, OW, C)`).
> 
> Say we omit `N` (which is typically 1) and distribute along `OH`, `OW`, and `C` (they are typically not 1). (This is actually how we distribute for pooling ops actually right now in IREE.) Then the `partitionedLoops` in the following current impl:
> 
> https://github.com/google/iree/blob/02cfcd1e5561c0db0470d09644a1f527c259c208/iree/compiler/Codegen/Utils/Utils.cpp#L134-L148
> 
> The `partitionedLoops` would return `[1, 2, 5]`. Then at L146 above, we could use `5` to index into `outputShape`, which only have 4 dimensions. That's OOB.
> 
> The real issue is that the current impl is assuming the first batch of implicit loops are used to indexing into the output. So the fix we need is to add the map from `loopToOutputDimMap` to break that false assumption. Avoid distributing along unit dims is something we certainly want, but it won't fix the OOB issue here. We also need the fix here. :)
> 
> I'm using the illformed `linalg.generic` as the test instead of using `linalg.pool` because we don't really have configurations for pooling ops as root to trigger this OOB; don't want to just add a configuration for this fix, as that has performance implications.
> 
> Does this make sense? Happy to explain more if not.

Thanks for the explanation.

**Review by MaheshRavishankar — APPROVED** [review]
